### ISSUE TITLE
[REQUEST] Implement CRR/AWAC

### ISSUE THREAD

**Issue by araffin** [takuseno/d3rlpy#4]
Hello,

Thanks for the resource.
It would be nice to implement [Critic Regularized Regression (CRR)](https://arxiv.org/abs/2006.15134) or [Advantage Weighted Actor Critic (AWAC)](https://arxiv.org/abs/2006.09359) (they share the same idea) and report the results on PyBullet envs.

Some implementations can be found in the [ACME repo](https://github.com/deepmind/acme/tree/master/acme/agents/tf/crr) or in [rlkit](https://github.com/vitchyr/rlkit/pull/115).
I also did a "quick and dirty" one in SB3 (dev branch), see [here](https://github.com/DLR-RM/stable-baselines3/blob/feat/crr/stable_baselines3/sac/sac.py#L392)

PS: sorry but I do not have the time to do a clean implementation myself... (already quite busy with SB3)

**Comment by takuseno** [comment]
@araffin Thank you for requesting an issue! I'm considering more algorithms based on Advantage-Weighted Critic which is already implemented in d3rlpy (not quite tested yet). So I'll update this issue when they are ready!

**Comment by takuseno** [comment]
@araffin I'm implementing AWAC algorithm on awac branch.
https://github.com/takuseno/d3rlpy/blob/awac/d3rlpy/algos/awac.py

When performance reports are available, I'll share them here.

**Comment by takuseno** [comment]
I was implementing AWAC based on the paper that says they made AWAC based on TD3. However, I realized that they did it based on SAC instead. For now, TD3-based AWAC performs good enough at offline training, but not good enough at finetuning. I'm suspecting that the reason is the exploration issue.

**Comment by araffin** [comment]
Thanks for the update =)

>However, I realized that they did it based on SAC instead. For now, TD3-based AWAC performs good enough at offline training, but not good enough at finetuning. I'm suspecting that the reason is the exploration issue.

You mean that what is written in the paper is different from what is implemented?

I also noticed some differences in your implementation. You use random actions to estimate the Value, where as they use the deterministic output of the policy for it.

EDIT: it seems that you updated that part but you are still sampling instead of using the mean/deterministic output

I also contacted the author recently and he told me that "the only notable thing is that we do not output the standard deviation (actually the log variance) of the policy with neural network layers, they are just learned parameters themselves (1 per action dimension). This prevents the variances from overfitting." (so like for PPO vs SAC)

**Comment by takuseno** [comment]
@araffin Thank you for check the update! I noticed the logstd parameter days ago. For the latest version, the performance is much better. I'll merge this in this week.

**Comment by araffin** [comment]
> @araffin Thank you for checking the update! I noticed the logstd parameter days ago. For the latest version, the performance is much better. I'll merge this in this week.

Good to hear =) (I'm been trying it too and it looks good)

minor remark, you may consider python [raw strings](https://www.journaldev.com/23598/python-raw-string) for the docstrings:

```python
r""":math:`\alpha/(\sqrt{v} + \epsilon)`"""
```

this avoids the use of too many backslashes ;)

**Comment by takuseno** [comment]
Oh, I did not know that! Thank you! I'll use this to remove doubled slashes.

**Comment by takuseno** [comment]
I've merged awac branch to master now. I've confirmed it somehow works, however, I did not thorough evaluation because it's a bit busy now.

**Comment by takuseno** [comment]
@araffin Thanks for your advice!
